src/extract/index.ts

```diff
@@ -0,0 +1,52 @@
+import { zodToJsonSchema } from 'openai-zod-to-json-schema';
+import { type z } from 'zod';
+
+import { type Model, type Msg, MsgUtil } from '../model/index.js';
+
+/**
+ * Extract data using OpenAI structured outputs.
+ *
+ * Always returns an object satisfying the provided Zod schema.
+ */
+export function createExtractFunction<Schema extends z.ZodObject<any>>(args: {
+  /** The ChatModel used to make API calls. */
+  chatModel: Model.Chat.Model;
+  /** A descriptive name for the object to extract. */
+  name: string;
+  /** The Zod schema for the data to extract. */
+  schema: Schema;
+  /** Add a system message to the beginning of the messages array. */
+  systemMessage: string;
+}): (input: string | Msg) => Promise<z.infer<Schema>> {
```

Comment:
would be nice to add a `strict?: boolean` flag here since it has a measurable impact on perf, especially for first-time runs. not sure whether the default should be `true` or `false`...